encoding/encoding.go

```diff
@@ -51,6 +52,20 @@ type Serializer interface {
 	CheckSuccess(body *transport.Response) error
 }
 
+// StreamRequestReader interface exposes method to read multiple request body
+type StreamRequestReader interface {
+	// NextBody returns the encoded request body if available, and if not, returns an
+	// io.EOF error
+	NextBody() ([]byte, error)
+}
+
+// StreamSerializer interface exposes additional methods to handle stream requests
+type StreamSerializer interface {
+	// StreamRequest creates a root stream request, a stream request reader using
+	// body reader provided
+	StreamRequest(body io.Reader) (*transport.Request, StreamRequestReader, error)
```

Comment:
One improvement we could do with this method is to have a `*transport.StreamRequest`, this will give the following benefits:
- No confusion on when to use the method `Request` or `StreamRequest`
- No confusion if `*transport.Request` is being used manually. You will one type for streaming and another for unary. The big benefits is that `*transport.StreamRequest` will not have a Body field, this will remove confusion about using `*transport.Request` with a stream request.
- No need to do the extra check in the methods `Request` and `StreamRequest`. You will not be able to use request with streaming anyway.

Thoughts?

Reply:
StreamRequest does make sense, but it's exactly replica of `transport.Request` except the body which would be nil in case of a streaming request. Also, it would again need bunch of duplicate implementation of certain request enricher and middleware watchers
https://github.com/yarpc/yab/blob/8f466890aa39477328191e0eccfdef4d04190eb2/request.go#L178

Transport reuses the `transport.Request` to transport to YARPC stream cleanly
https://github.com/yarpc/yab/blob/f84a2941e8375ef8f980a41aeefa650059af1b0d/transport/grpc.go#L146

I believe this would overwhelm the benefits

Reply:
sync offline - @jronak agreed to have StreamRequest